src/gui/SampleWaveform.cpp

```diff
@@ -23,73 +23,93 @@
  */
 
 #include "SampleWaveform.h"
+#include "interpolation.h"
 
 namespace lmms::gui {
 
-void SampleWaveform::visualize(Parameters parameters, QPainter& painter, const QRect& rect)
+SampleWaveform::SampleWaveform(const SampleFrame* buffer, size_t size)
+	: m_buffer(buffer)
+	, m_size(size)
 {
-	const int x = rect.x();
-	const int height = rect.height();
-	const int width = rect.width();
-	const int centerY = rect.center().y();
+}
 
-	const int halfHeight = height / 2;
+void SampleWaveform::generate()
+{
+	if (!m_buffer || m_size == 0) { return; }
 
-	const auto color = painter.pen().color();
-	const auto rmsColor = color.lighter(123);
+	m_min.clear();
+	m_max.clear();
 
-	const float framesPerPixel = std::max(1.0f, static_cast<float>(parameters.size) / width);
+	const auto maxLevel = static_cast<int>(std::log2(m_size));
+	for (auto level = 1; level <= maxLevel; ++level)
+	{
+		const auto downsampledSize = m_size / static_cast<int>(std::exp2(level));
+		m_min[level].resize(downsampledSize);
+		m_max[level].resize(downsampledSize);
+
+		if (level == 1)
+		{
+			for (auto i = std::size_t{0}; i < downsampledSize; ++i)
+			{
+				static auto peakComp = [](const SampleFrame& a, const SampleFrame& b) { return a.average() < b.average(); };
+				m_min[level][i] = std::min_element(&m_buffer[i * 2], &m_buffer[i * 2 + 2], peakComp)->average();
+				m_max[level][i] = std::max_element(&m_buffer[i * 2], &m_buffer[i * 2 + 2], peakComp)->average();
+			}
+		}
+		else
+		{
+			for (auto i = std::size_t{0}; i < downsampledSize; ++i)
+			{
+				m_min[level][i] = *std::min_element(&m_min[level - 1][i * 2], &m_min[level - 1][i * 2 + 2]);
+				m_max[level][i] = *std::max_element(&m_max[level - 1][i * 2], &m_max[level - 1][i * 2 + 2]);
+			}
+		}
+	}
+}
 
-	constexpr float maxFramesPerPixel = 512.0f;
-	const float resolution = std::max(1.0f, framesPerPixel / maxFramesPerPixel);
-	const float framesPerResolution = framesPerPixel / resolution;
+void SampleWaveform::visualize(QPainter& painter, const QRect& rect, float amplification, bool reversed,
+	std::optional<size_t> from, std::optional<size_t> to)
+{
+	if (!m_buffer || m_size == 0) { return; }
 
-	size_t numPixels = std::min(parameters.size, static_cast<size_t>(width));
-	auto min = std::vector<float>(numPixels, 1);
-	auto max = std::vector<float>(numPixels, -1);
-	auto squared = std::vector<float>(numPixels, 0);
+	const auto sampleBegin = from.has_value() ? from.value() : 0;
+	const auto sampleEnd = to.has_value() ? to.value() : m_size;
```

Comment:
```suggestion
	const auto sampleBegin = from.value_or(0);
	const auto sampleEnd = to.value_or(m_size);
```

Reply:
Oh yeah, I told myself I was going to make this change but I completely forgot, thanks.